### ISSUE TITLE
Behind a reverse proxy

### ISSUE THREAD

**Issue by stappersg** [tweedegolf/mailcrab#55]
Summary: What to do extra / special to put `mailcrab` behind a _reverse proxy_

**Just** `mailcarb` works as documented in the `README.md`. ( backend: `cargo run`, frontend: `trunk serve`, sendmail: `swaks --to foo@bar --server 127.0.0.1:1025`, URL: http://127.0.0.1:1080 )

**Behind** a _reverse proxy_ I have not so much success.

It is _Apache_ that I use as _reverse proxy_, this is in `/etc/apache2/sites-available/000-default.conf`:
```text
$ grep -ve \# /etc/apache2/sites-available/000-default.conf
<VirtualHost *:80>

	ServerAdmin webmaster@localhost
	DocumentRoot /var/www/html

	<location /mailcrab>
        Require all granted
        ProxyPass http://127.0.0.1:1080
        ProxyPassReverse http://127.0.0.1:1080
        </location>


	ErrorLog ${APACHE_LOG_DIR}/error.log
	CustomLog ${APACHE_LOG_DIR}/access.log combined

</VirtualHost>
$
```

Visiting URL http://127.0.0.1/mailcrab or http://127.0.0.1/mailcrab/ gives **nothing**.
Visiting URL http://127.0.0.1/mailcrab/api/messages gives **JSON**.

Webbrowser is firefox-esr, 91.12.0esr-1~deb10u1

**Comment by stappersg** [comment]
Intermediate report:  It seems to be webbrowser related.

Visiting http://localhost/mailcrab with webbrowser yields an empty page. With `tcpdump -i lo port 1080` can HTTP error 404 been seen.

The same URL with `curl` yields the `index.html`
```text
$ curl --silent http://127.0.0.1/mailcrab | grep html
<!DOCTYPE html><html lang="en"><head>
</script></body></html>
$
```

**Comment by marlonbaeten** [comment]
Are you setting `MAILCRAB_PREFIX=mailcrab` in the environment?

**Comment by marlonbaeten** [comment]
Running a websocket connection over a proxy is requires more than just proxying the HTTP traffic. See for instance https://httpd.apache.org/docs/2.4/mod/mod_proxy_wstunnel.html

**Comment by stappersg** [comment]
_Acknowledge_ I will dive further into this. For now:
```text
paddy:~# dpkg -l apache2 | grep ^ii
ii  apache2        2.4.51-2     amd64        Apache HTTP Server
paddy:~# a2query -m | grep proxy
proxy_http (enabled by site administrator)
proxy (enabled by site administrator)
paddy:~# 
```

**Comment by stappersg** [comment]
On Fri, Apr 28, 2023 at 07:19:34AM -0700, Geert Stappers wrote:
> I will dive further into this.

I have succes with

|$ grep -v -e \# /etc/apache2/sites-available/000-default.conf
|<VirtualHost *:80>
|
|	ServerAdmin ***@***.***
|	DocumentRoot /var/www/html
|
|	ProxyPass /mailcrab/ws "ws://127.0.0.1:1080/ws"
|       ProxyPassReverse /mailcrab/ws "ws://127.0.0.1:1080/ws"
|
|       ProxyPass /mailcrab "http://127.0.0.1:1080"
|       ProxyPassReverse /mailcrab "http://127.0.0.1:1080"
|
|	ErrorLog ${APACHE_LOG_DIR}/error.log
|	CustomLog ${APACHE_LOG_DIR}/access.log combined
|
|</VirtualHost>
|$

That is on another system, it has Apache2 2.4.57-2.
Testing it with Apache2 2.4.51-2 is for later.

Summary for now:
  * MAILCRAB_PREFIX
  * proxy websocket


Groeten
Geert Stappers
-- 
Silence is hard to parse

**Comment by stappersg** [comment]
Yeah, works also on Apache 2.4.51-2 with `mod_proxy` and `mod_proxy_http`.

For a _vhost_ I got succes with:
```text
$ grep -v \# /etc/apache2/sites-available/mailcrab.vincent.mig.conf 
<VirtualHost *:80>
        ServerAdmin stappers@stappers.it
        ServerName mailcrab.vincent.mig

        Documentroot /dev/null

        ProxyPass /ws "ws://127.0.0.1:1080/ws"
        ProxyPassReverse /ws "ws://127.0.0.1:1080/ws"

        ProxyPass / "http://127.0.0.1:1080/"
        ProxyPassReverse / "http://127.0.0.1:1080/"

        ErrorLog ${APACHE_LOG_DIR}/error.log

        LogLevel warn

        CustomLog ${APACHE_LOG_DIR}/mailcrab.vincent.mig.access.log combined

</VirtualHost>
```

Q: What to do extra / special to put `mailcrab` behind a _reverse proxy_?
A: For Apache: Do also proxy the websocket, `"ws://127.0.0.1:1080/ws"`. For all,
using Path prefix, use environment variable `MAILCRAB_PREFIX`.

(Issue is _solved_ and left open with a next question: Where to document "Beware, websockets!"? )

**Comment by marlonbaeten** [comment]
We could add a section to the README about proxying the mailcrab frontend if we keep it brief. Would be nice to also include a hint on how to proxy websockets with nginx. Any PR's are welcome on this.
